moztelemetry/stats.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from __future__ import division
+
+import math
+from collections import namedtuple
+
+
+def rank(sample):
+    rank = 1
+    ranks = {}
+
+    for k in sorted(sample.keys()):
+        n = sample[k]
+        ranks[k] = rank + (n - 1) / 2
+        rank += n
+
+    return ranks
+
+
+def tie_correct(sample):
+    tc = 0
+    n = sum(sample.values())
+
+    if n == 1:
+        return 1.0  # Avoid a ``ZeroDivisionError``.
+
+    for k in sorted(sample.keys()):
+        tc += math.pow(sample[k], 3) - sample[k]
+    tc = 1 - tc / (math.pow(n, 3) - n)
+
+    return tc
+
+
+def ndtr(v):
+    if not v:
+        return v
+
+    sqrth = math.sqrt(2) / 2
+    x = v * sqrth
+    z = abs(x)
+    if z < sqrth:
+        y = 0.5 + 0.5 * math.erf(x)
+    else:
+        y = 0.5 * math.erfc(z)
+    if x > 0:
+        y = 1 - y
+    return y
+
+
+mwu_result = namedtuple('Mann_Whitney_U', ('u', 'p'))
+
+
+def mann_whitney_u(sample1, sample2, use_continuity=True):
+    """
+    Computes the Mann-Whitney rank test on both samples.
+
+    Each sample is expected to be of the form::
+
+        {1: 5, 2: 20, 3: 12, ...}
+
+    Returns the U statistic, equal to min(U for sample1, U for sample2).
+
+    """
+    # Merge dictionaries, adding values if keys match.
+    sample = sample1.copy()
+    for k, v in sample2.items():
+        sample[k] = sample.get(k, 0) + v
+
+    # Create a ranking dictionary using same keys for lookups.
+    ranks = rank(sample)
+
+    # Calculate Mann-Whitney U for both samples.
+    sum_of_ranks = sum([sample1[k] * ranks[k] for k, v in sample1.items()])
+
+    n1 = sum(sample1.values())
+    n2 = sum(sample2.values())
+
+    u1 = sum_of_ranks - (n1 * (n1 + 1)) / 2
+    u2 = n1 * n2 - u1
+
+    tie_correction = tie_correct(sample)
+    sd_u = math.sqrt(tie_correction * n1 * n2 * (n1 + n2 + 1) / 12.0)
+    mean_rank = n1 * n2 / 2.0 + 0.5 * use_continuity
+    z = abs((max(u1, u2) - mean_rank) / sd_u)
```

Comment:
This should be the `min(u1, u2)`, [see here](http://psych.unl.edu/psycrs/handcomp/hcmann.PDF)

Reply:
In step 10 they're using U=32 to compute `z`, which is the `max(Ua, Ub)` in the example.